src/layout/Toolbar.js

```diff
@@ -0,0 +1,103 @@
+    Toolbar.prototype.publish("borderStyle", "solid", "string", "Toolbar Height",null,{tags:['basic']});
+    Toolbar.prototype.publish("borderRadius", 0, "number", "Border Radius",null,{tags:['basic']});
+    Toolbar.prototype.publish("backgroundColor", null, "html-color", "Background Color",null,{tags:['basic']});
+
+    Toolbar.prototype.toolbarAnnotations = function(_){
+        if (!arguments.length) { return this._tabAnnotations; }
+        this._tabAnnotations = _;
+        return this;
+    }
+
+    Toolbar.prototype.testData = function () {
+        this.toolbarAnnotations([
+            {
+                width:55,
+                height: 25,
+                type: "button",
+                widget: new formInput().type("button").value("button 1").label("button 1").name("button1")
+            },
+            {
+                width:55,
+                height: 25,
+                type: "button",
+                widget: new formInput().type("button").value("button 2").label("button 2").name("button2")
+            },
+            {
+                width:25,
+                height: 25,
+                type: "icon",
+                widget: new Icon().testData()
+            }
+        ]);
+        return this;
+    };
+
+    Toolbar.prototype.enter = function (domNode, element) {
+        HTMLWidget.prototype.enter.apply(this, arguments);
+        this._toolbarContainer = element.append("div").attr("class", "toolbar-container");
+    };
+
+    Toolbar.prototype.update = function (domNode, element) {
+        HTMLWidget.prototype.update.apply(this, arguments);
+        var context = this;
+
+        this._toolbarContainer
+            .style("border-width",this.borderWidth()+"px")
+            .style("border-color",this.borderColor())
+            .style("border-style",this.borderStyle())
+            .style("border-radius",this.borderRadius()+"px")
+            .style("background-color",this.backgroundColor())
+        ;
+
+        var widgets = this._toolbarContainer.selectAll(".toolbar-widget").data(this.toolbarAnnotations());
+
+        widgets.enter().append("div")
+            .attr("class", "toolbar-widget")
+            .each(function (obj, idx) {
+                if (obj.type !== "button") {
+                    d3.select(this).style("width",obj.width+"px");
+                    d3.select(this).style("height",obj.height+"px");
+                    obj.widget.target(this);
+                }
+                if (obj.type === "button") {
+                    obj.widget.target(this).render(function(widget) {
+                        widget._inputElement.style("display","inline-block");
+                        widget._element.style("width",obj.width+"px");
+                        widget._inputElement.style("height",obj.height+"px");
+                    });
+                }
+                d3.select(this).style("padding-left",(idx > 0 ? (context.gutter()/2)+"px" : null));
```

Comment:
These publish parameters should be applied outside of the context of .enter() so that they'll be applied whenever these publish param values change (instead of solely applying on the initial rendering of each data node).